# Part 0 - (Dataset Wrangling Airline On-Time Performance Data)
## by (Lariane Mohcene Mouad)

## Introduction
> **Introducing the dataset** 	
This dataset reports flights in the United States in 2007, including carriers, arrival and departure delays, and reasons for delays and many other variables

**Variables Description**
- **Year**: 1987 - 2008 (since we're studying only flight_2007 dataset, all years will be set to 2007)
- **Month**: 1 (Jan) - 12 (Dec)
- **DayofMonth**: 1 - 31
- **DayOfWeek**: 1 (Monday) - 7 (Sunday)
- **DepTime**: actual departure time (local, hhmm)
- **ArrTime**: actual arrival time (local, hhmm)
- **CRSDepTime**: scheduled departure time (local, hhmm)
- **CRSArrTime**: scheduled arrival time (local, hhmm)
    - **CRS**: Computer Reservation System. CRS provide information on airline schedules, fares and seat availability to travel agencies and allow agents to book seats and issue tickets.
- **UniqueCarrier**: unique carrier code which is the Carrier Code most recently used by a carrier. A numeric suffix is used to distinguish duplicate codes, for example, PA, PA (1), PA (2). Use this field to perform analysis of data reported by one and only one carrier.
- **FlightNum**: flight number
- **TailNum**: plane tail number
- **ActualElapsedTime**: The time computed from gate departure time to gate arrival time in minutes
- **CRSElapsedTime**: scheduled elapsedTime in minutes
- **AirTime**: Flight Time, in Minutes
- **ArrDelay**: Difference in minutes between scheduled and actual arrival time. Early arrivals show negative numbers.
- **DepDelay**: Difference in minutes between scheduled and actual departure time. Early departures show negative numbers.
- **Origin**: origin IATA airport code
- **Dest**: destination IATA airport code
- **Distance**: The distance between the Origin and the destination in miles
- **TaxiIn**: taxi in time which is the time elapsed between wheels down and arrival at the destination airport gate in minutes
- **TaxiOut**: taxi out time which the time elapsed between departure from the origin airport gate and wheels off in minutes
- **Cancelled**: was the flight cancelled? "1 = yes, 0 = no"
- **CancellationCode**: "reason for cancellation (A = carrier, B = weather, C = NAS, D = security)" 
    - **NAS** : National Aviation System is a broad term for a set of situations that can affect flight times. These include airport operations, non-risky weather conditions, high air traffic volume, and air traffic control delays. For unknown reasons.
    - **Weather**: Flight cancellation due to weather conditions
    - **security**: it’s a result of terminal or concourse evacuation, a security breach on an aircraft, long lines at screening areas, or defective screening equipment
- **Diverted**: A flight that is required to land at a destination other than the original scheduled destination for reasons beyond the control of the pilot/company "1 = yes, 0 = no"
- **CarrierDelay**: in minutes, usually pertains to the status within the airline's control. For example, problems with maintenance and crew, cleaning within the cabin, fueling, and baggage loading could all be contributing factors of a delayed flight 
- **WeatherDelay**: in minutes
- **NASDelay**: in minutes
- **SecurityDelay**: in minutes
- **LateAircraftDelay**: in minutes

**References:**
- [Different Types of Flight Delays](https://www.sheffield.com/2019/different-flight-delays.html#:~:text=of%20flight%20delays.-,Air%20Carrier,factors%20of%20a%20delayed%20flight.)    
- [Dataset official website](https://www.transtats.bts.gov/DatabaseInfo.asp?QO_VQ=EFD&Yv0x=D)
- [Definitions](https://www.transtats.bts.gov/Fields.asp?gnoyr_VQ=FGJ)


In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb
import datetime as dt

%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [2]:
# Setting pandas option to display all columns
pd.set_option('display.max_columns', None)

In [3]:
# Loading and reading the dataset
flights = pd.read_csv('2007.csv')

In [4]:
flights.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2007      1           1          1   1232.0        1225   1341.0   
1  2007      1           1          1   1918.0        1905   2043.0   
2  2007      1           1          1   2206.0        2130   2334.0   
3  2007      1           1          1   1230.0        1200   1356.0   
4  2007      1           1          1    831.0         830    957.0   

   CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0        1340            WN       2891    N351               69.0   
1        2035            WN        462    N370               85.0   
2        2300            WN       1229    N685               88.0   
3        1330            WN       1355    N364               86.0   
4        1000            WN       2278    N480               86.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0            75.0     54.0       1.0       7.0    SMF  ONT       389       4   
1            90.0     74.0       8.0      13.0    SMF  PDX       479       5   
2            90.0     73.0      34.0      36.0    SMF  PDX       479       6   
3            90.0     75.0      26.0      30.0    SMF  PDX       479       3   
4            90.0     74.0      -3.0       1.0    SMF  PDX       479       3   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0       11          0              NaN         0             0             0   
1        6          0              NaN         0             0             0   
2        9          0              NaN         0             3             0   
3        8          0              NaN         0            23             0   
4        9          0              NaN         0             0             0   

   NASDelay  SecurityDelay  LateAircraftDelay  
0         0              0                  0  
1         0              0                  0  
2         0              0                 31  
3         0              0                  3  
4         0              0                  0

In [5]:
# high-level overview of data shape and composition
print(flights.shape)
print(flights.dtypes)

(7453215, 29)
Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn                 int64
TaxiOut                int64
Cancelled              int64
CancellationCode      object
Diverted               int64
CarrierDelay           int64
WeatherDelay           int64
NASDelay               int64
SecurityDelay          int64
LateAircraftDelay      int64
dtype: object


### What is the structure of your dataset?

> There are 7453215 flight records in the dataset with 29 features including carriers, arrival and departure delays, and reasons for delays and many other variables. Most variables are numeric in nature, and the rest of them are none-numeric. 


### What is/are the main feature(s) of interest in your dataset?

> What and when flights are more likely to be cancelled or delayed? and what are the reasons behind it? and is there any destination or arrival cities that are home to more delays or cancellations? So the mains features would be all features related to cancelation and delay and also features such as distance, actual/scheduled elapsed time, arrDelay/Depdelay.

### What features in the dataset do you think will help support your investigation into your feature(s) of interest?

> I expect that the flight month and the flight Day of Week, Day of Month will have huge effect on whether or not the flight will be cancelled or delayed : flights during winter months are more likely to be cancelled because of the possibility of risky weather conditions. also the flight Day Of Week and flight day of month (a k a flight date) can also effect this feature where weekends and special occasions are more likely to have a high air traffic volume, and air traffic control delays which can lead to flight cancelation and delay. Also origin, dest features would be helpful

### Selecting a random sample from the dataset to work on 
Since the original dataset is too large ( + 7m rows ), I will do my study on only a random 1.5m records sample in order to reduce the execution time.

In [6]:
flights = flights.sample(n=1500000)
flights.reset_index(drop=True, inplace=True)

## Preliminary Wrangling

### Assessing Data

**Quality issues**:
- ``DepTime`` ``CRSDepTime`` ``ArrTime`` ``CRSArrTime`` fix these features values time format
- Replace DayOfWeek values by the actual days names
- Replace the bool features values by their actual values

**Tidiness issues**:
- ``Year`` ``Month`` ``DayofMonth`` should be in one column named ``Date``

### Data Cleaning


In [7]:
# Making a copy 
df = flights.copy()

### Quality issues

#### Define
- Convert ``DepTime`` ``CRSDepTime`` ``ArrTime`` ``CRSArrTime`` values to string.
- Remove **".0"** at the end
- Change values format to **hh:mm:ss** by ``DepTime`` ``CRSDepTime`` ``ArrTime`` ``CRSArrTime`` type to datetime.time

#### Code

In [8]:
# List of features to deal with
time_vars = ['DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime']

# Convert features values to string.
for var in time_vars:
    df[var] = df[var].astype(str)
    
# Remove **".0"** at the end
for var in ['DepTime', 'ArrTime']:    
    df[var] = df[var].str[:-2]

In [11]:
df.head()

Year  Month  DayofMonth  DayOfWeek DepTime CRSDepTime ArrTime CRSArrTime  \
0  2007      6          24          7    1151       1150    1313       1327   
1  2007      5           7          1    1813       1815    1844       1854   
2  2007     12           2          7    1746       1755    1816       1824   
3  2007      4          22          7    1630       1620    1753       1750   
4  2007      8           4          6    1908       1910    2035       2040   

  UniqueCarrier  FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  \
0            XE       2773  N12946               82.0            97.0   
1            OO       6154  N217SW               31.0            39.0   
2            AQ         48  N841AL               30.0            29.0   
3            OH       4983  N442CA               83.0            90.0   
4            DL         65   N6702               87.0            90.0   

   AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  TaxiOut  \
0     62.0     -14.0       1.0    EWR  GSO       446       4       16   
1     16.0     -10.0      -2.0    LAX  SNA        36       4       11   
2     24.0      -8.0      -9.0    LIH  HNL       102       2        4   
3     53.0       3.0      10.0    JFK  DCA       213       3       27   
4     57.0      -5.0      -2.0    ATL  TPA       406       4       26   

   Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  NASDelay  \
0          0              NaN         0             0             0         0   
1          0              NaN         0             0             0         0   
2          0              NaN         0             0             0         0   
3          0              NaN         0             0             0         0   
4          0              NaN         0             0             0         0   

   SecurityDelay  LateAircraftDelay  
0              0                  0  
1              0                  0  
2              0                  0  
3              0                  0  
4              0                  0

In [41]:
# Creating function to fix values format and type
def hhmm(df, var):
    v = df[var]
    if len(v) == 4:
        mm, hh = int(float(v[2:])), int(float(v[:-2]))
    elif len(v) == 3:
        mm, hh = int(float(v[1:])), int(float(v[:-2]))
    else: 
        mm, hh = 0, 0
    wrong = (hh not in range(24)) or (hh not in range(59))
    if wrong:
        return dt.time(0, 0)
    else:
        return dt.time(hh, mm)

In [40]:
# Applying the function
for var in time_vars:
    df[var] = df.apply(hhmm, args=(var,), axis=1)

#### Test

In [42]:
# Checking Values format
df.head()

Year  Month  DayofMonth  DayOfWeek   DepTime CRSDepTime   ArrTime  \
0  2007      6          24          7  11:51:00   11:50:00  13:13:00   
1  2007      5           7          1  18:13:00   18:15:00  18:44:00   
2  2007     12           2          7  17:46:00   17:55:00  18:16:00   
3  2007      4          22          7  16:30:00   16:20:00  17:53:00   
4  2007      8           4          6  19:08:00   19:10:00  20:35:00   

  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0   13:27:00            XE       2773  N12946               82.0   
1   18:54:00            OO       6154  N217SW               31.0   
2   18:24:00            AQ         48  N841AL               30.0   
3   17:50:00            OH       4983  N442CA               83.0   
4   20:40:00            DL         65   N6702               87.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0            97.0     62.0     -14.0       1.0    EWR  GSO       446       4   
1            39.0     16.0     -10.0      -2.0    LAX  SNA        36       4   
2            29.0     24.0      -8.0      -9.0    LIH  HNL       102       2   
3            90.0     53.0       3.0      10.0    JFK  DCA       213       3   
4            90.0     57.0      -5.0      -2.0    ATL  TPA       406       4   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0       16          0              NaN         0             0             0   
1       11          0              NaN         0             0             0   
2        4          0              NaN         0             0             0   
3       27          0              NaN         0             0             0   
4       26          0              NaN         0             0             0   

   NASDelay  SecurityDelay  LateAircraftDelay  
0         0              0                  0  
1         0              0                  0  
2         0              0                  0  
3         0              0                  0  
4         0              0                  0

In [43]:
# Checking Values type
df.iloc[0].DepTime

datetime.time(11, 51)

#### Define
- Replace DayOfWeek values by the actual days names

#### Code

In [44]:
day_of_week = {
1: 'Monday',
2: 'Tuesday',
3: 'Wednesday',
4: 'Thursday',
5: 'Friday',
6: 'saturday',
7: 'Sunday'
}
df.DayOfWeek.replace(day_of_week, inplace=True)

#### Test

In [45]:
df.head()

Year  Month  DayofMonth DayOfWeek   DepTime CRSDepTime   ArrTime  \
0  2007      6          24    Sunday  11:51:00   11:50:00  13:13:00   
1  2007      5           7    Monday  18:13:00   18:15:00  18:44:00   
2  2007     12           2    Sunday  17:46:00   17:55:00  18:16:00   
3  2007      4          22    Sunday  16:30:00   16:20:00  17:53:00   
4  2007      8           4  saturday  19:08:00   19:10:00  20:35:00   

  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0   13:27:00            XE       2773  N12946               82.0   
1   18:54:00            OO       6154  N217SW               31.0   
2   18:24:00            AQ         48  N841AL               30.0   
3   17:50:00            OH       4983  N442CA               83.0   
4   20:40:00            DL         65   N6702               87.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0            97.0     62.0     -14.0       1.0    EWR  GSO       446       4   
1            39.0     16.0     -10.0      -2.0    LAX  SNA        36       4   
2            29.0     24.0      -8.0      -9.0    LIH  HNL       102       2   
3            90.0     53.0       3.0      10.0    JFK  DCA       213       3   
4            90.0     57.0      -5.0      -2.0    ATL  TPA       406       4   

   TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0       16          0              NaN         0             0             0   
1       11          0              NaN         0             0             0   
2        4          0              NaN         0             0             0   
3       27          0              NaN         0             0             0   
4       26          0              NaN         0             0             0   

   NASDelay  SecurityDelay  LateAircraftDelay  
0         0              0                  0  
1         0              0                  0  
2         0              0                  0  
3         0              0                  0  
4         0              0                  0

#### Define
- Replace the bool features values by their actual values, 1 = yes, 0 = No

#### Code

In [46]:
bool_values = {1: 'Yes', 0:'No'}
df.Cancelled.replace(bool_values, inplace=True)
df.Diverted.replace(bool_values, inplace=True)

#### Test

In [47]:
df.head()

Year  Month  DayofMonth DayOfWeek   DepTime CRSDepTime   ArrTime  \
0  2007      6          24    Sunday  11:51:00   11:50:00  13:13:00   
1  2007      5           7    Monday  18:13:00   18:15:00  18:44:00   
2  2007     12           2    Sunday  17:46:00   17:55:00  18:16:00   
3  2007      4          22    Sunday  16:30:00   16:20:00  17:53:00   
4  2007      8           4  saturday  19:08:00   19:10:00  20:35:00   

  CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0   13:27:00            XE       2773  N12946               82.0   
1   18:54:00            OO       6154  N217SW               31.0   
2   18:24:00            AQ         48  N841AL               30.0   
3   17:50:00            OH       4983  N442CA               83.0   
4   20:40:00            DL         65   N6702               87.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0            97.0     62.0     -14.0       1.0    EWR  GSO       446       4   
1            39.0     16.0     -10.0      -2.0    LAX  SNA        36       4   
2            29.0     24.0      -8.0      -9.0    LIH  HNL       102       2   
3            90.0     53.0       3.0      10.0    JFK  DCA       213       3   
4            90.0     57.0      -5.0      -2.0    ATL  TPA       406       4   

   TaxiOut Cancelled CancellationCode Diverted  CarrierDelay  WeatherDelay  \
0       16         0              NaN        0             0             0   
1       11         0              NaN        0             0             0   
2        4         0              NaN        0             0             0   
3       27         0              NaN        0             0             0   
4       26         0              NaN        0             0             0   

   NASDelay  SecurityDelay  LateAircraftDelay  
0         0              0                  0  
1         0              0                  0  
2         0              0                  0  
3         0              0                  0  
4         0              0                  0

### Tidiness Issues

#### Define:
- ``Year`` ``Month`` ``DayofMonth`` should be in one column named ``Date``

#### Code

In [48]:
# Creating date column from Year, Month, DayofMonth
df['Date'] = pd.to_datetime(dict(year=df.Year, month=df.Month, day=df.DayofMonth))

In [49]:
# Dropping columns Year, Month, DayofMonth
df.drop(['Year', 'Month', 'DayofMonth'], axis=1, inplace=True)

#### Test

In [50]:
df.head()

DayOfWeek   DepTime CRSDepTime   ArrTime CRSArrTime UniqueCarrier  \
0    Sunday  11:51:00   11:50:00  13:13:00   13:27:00            XE   
1    Monday  18:13:00   18:15:00  18:44:00   18:54:00            OO   
2    Sunday  17:46:00   17:55:00  18:16:00   18:24:00            AQ   
3    Sunday  16:30:00   16:20:00  17:53:00   17:50:00            OH   
4  saturday  19:08:00   19:10:00  20:35:00   20:40:00            DL   

   FlightNum TailNum  ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  \
0       2773  N12946               82.0            97.0     62.0     -14.0   
1       6154  N217SW               31.0            39.0     16.0     -10.0   
2         48  N841AL               30.0            29.0     24.0      -8.0   
3       4983  N442CA               83.0            90.0     53.0       3.0   
4         65   N6702               87.0            90.0     57.0      -5.0   

   DepDelay Origin Dest  Distance  TaxiIn  TaxiOut Cancelled CancellationCode  \
0       1.0    EWR  GSO       446       4       16         0              NaN   
1      -2.0    LAX  SNA        36       4       11         0              NaN   
2      -9.0    LIH  HNL       102       2        4         0              NaN   
3      10.0    JFK  DCA       213       3       27         0              NaN   
4      -2.0    ATL  TPA       406       4       26         0              NaN   

  Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0        0             0             0         0              0   
1        0             0             0         0              0   
2        0             0             0         0              0   
3        0             0             0         0              0   
4        0             0             0         0              0   

   LateAircraftDelay       Date  
0                  0 2007-06-24  
1                  0 2007-05-07  
2                  0 2007-12-02  
3                  0 2007-04-22  
4                  0 2007-08-04

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1500000 entries, 0 to 1499999
Data columns (total 27 columns):
 #   Column             Non-Null Count    Dtype         
---  ------             --------------    -----         
 0   DayOfWeek          1500000 non-null  object        
 1   DepTime            1500000 non-null  object        
 2   CRSDepTime         1500000 non-null  object        
 3   ArrTime            1500000 non-null  object        
 4   CRSArrTime         1500000 non-null  object        
 5   UniqueCarrier      1500000 non-null  object        
 6   FlightNum          1500000 non-null  int64         
 7   TailNum            1499997 non-null  object        
 8   ActualElapsedTime  1464011 non-null  float64       
 9   CRSElapsedTime     1499807 non-null  float64       
 10  AirTime            1464011 non-null  float64       
 11  ArrDelay           1464011 non-null  float64       
 12  DepDelay           1467489 non-null  float64       
 13  Origin             1500000 

### Ordering features

#### Code

In [53]:
df = df[['Date', 'DayOfWeek', 'UniqueCarrier', 'FlightNum', 'TailNum', 
         'Origin', 'Dest', 'Distance', 'DepTime', 'CRSDepTime', 'ArrTime', 'CRSArrTime', 
        'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay', 'DepDelay', 
        'TaxiIn', 'TaxiOut', 
        'Diverted', 'Cancelled', 'CancellationCode', 
        'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']]

#### Test

In [54]:
df.head()

Date DayOfWeek UniqueCarrier  FlightNum TailNum Origin Dest  Distance  \
0 2007-06-24    Sunday            XE       2773  N12946    EWR  GSO       446   
1 2007-05-07    Monday            OO       6154  N217SW    LAX  SNA        36   
2 2007-12-02    Sunday            AQ         48  N841AL    LIH  HNL       102   
3 2007-04-22    Sunday            OH       4983  N442CA    JFK  DCA       213   
4 2007-08-04  saturday            DL         65   N6702    ATL  TPA       406   

    DepTime CRSDepTime   ArrTime CRSArrTime  ActualElapsedTime  \
0  11:51:00   11:50:00  13:13:00   13:27:00               82.0   
1  18:13:00   18:15:00  18:44:00   18:54:00               31.0   
2  17:46:00   17:55:00  18:16:00   18:24:00               30.0   
3  16:30:00   16:20:00  17:53:00   17:50:00               83.0   
4  19:08:00   19:10:00  20:35:00   20:40:00               87.0   

   CRSElapsedTime  AirTime  ArrDelay  DepDelay  TaxiIn  TaxiOut Diverted  \
0            97.0     62.0     -14.0       1.0       4       16        0   
1            39.0     16.0     -10.0      -2.0       4       11        0   
2            29.0     24.0      -8.0      -9.0       2        4        0   
3            90.0     53.0       3.0      10.0       3       27        0   
4            90.0     57.0      -5.0      -2.0       4       26        0   

  Cancelled CancellationCode  CarrierDelay  WeatherDelay  NASDelay  \
0         0              NaN             0             0         0   
1         0              NaN             0             0         0   
2         0              NaN             0             0         0   
3         0              NaN             0             0         0   
4         0              NaN             0             0         0   

   SecurityDelay  LateAircraftDelay  
0              0                  0  
1              0                  0  
2              0                  0  
3              0                  0  
4              0                  0

### Saving cleaned dataset

In [55]:
df.to_csv('clean_flights_sample.csv', index=False)